`java.lang.IndexOutOfBoundsException: Index: -1` in indexing org.scala-lang/scala-library/2.13.8

Bug surfaced while auto-indexing https://sourcegraph.com/maven/org.scala-lang/scala-library/-/code-graph/indexes/TFNJRkluZGV4OjI0NDE3Mzk=, exception throwing for the following symbol https://sourcegraph.com/maven/org.scala-lang/scala-library/-/blob/scala/Function6.scala?L30

```
+ scip-java index --build-tool=scip
Using JVM version '8'
Compiling 32 Java sources
$ /root/.cache/coursier/jvm/adopt@1.8.0-292/bin/javac @/data/target/javacopts.txt
Generating SCIP...com.sourcegraph.scip_semanticdb.SignatureFormatterException: failed to format symbol 'scala/Function6#curried().'
symbol: "scala/Function6#curried()."
kind: METHOD
display_name: "curried"
annotations {
  tpe {
    type_ref {
      symbol: "scala/annotation/unspecialized#"
    }
  }
}
language: SCALA
signature {
  method_signature {
    type_parameters {
    }
    return_type {
      type_ref {
        symbol: "scala/Function1#"
        type_arguments {
          type_ref {
            symbol: "scala/Function6#[T1]"
          }
        }
        type_arguments {
          type_ref {
            symbol: "scala/Function1#"
            type_arguments {
              type_ref {
                symbol: "scala/Function6#[T2]"
              }
            }
            type_arguments {
              type_ref {
                symbol: "scala/Function1#"
                type_arguments {
                  type_ref {
                    symbol: "scala/Function6#[T3]"
                  }
                }
                type_arguments {
                  type_ref {
                    symbol: "scala/Function1#"
                    type_arguments {
                      type_ref {
                        symbol: "scala/Function6#[T4]"
                      }
                    }
                    type_arguments {
                      type_ref {
                        symbol: "scala/Function1#"
                        type_arguments {
                          type_ref {
                            symbol: "scala/Function6#[T5]"
                          }
                        }
                        type_arguments {
                          type_ref {
                            symbol: "scala/Function1#"
                            type_arguments {
                              type_ref {
                                symbol: "scala/Function6#[T6]"
                              }
                            }
                            type_arguments {
                              type_ref {
                                symbol: "scala/Function6#[R]"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
access {
  public_access {
  }
}
Caused by: java.lang.IndexOutOfBoundsException: Index: -1
	at java.util.Collections$EmptyList.get(Collections.java:4456)
	at com.sourcegraph.semanticdb_javac.Semanticdb$TypeRef.getTypeArguments(Semanticdb.java:23534)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatType(SignatureFormatter.java:573)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.SliceOps$1$1.accept(SliceOps.java:204)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1361)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:499)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:486)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatType(SignatureFormatter.java:570)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatScalaMethodSignature(SignatureFormatter.java:283)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatMethodSignature(SignatureFormatter.java:211)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbolUnsafe(SignatureFormatter.java:67)
	at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbol(SignatureFormatter.java:56)
	at com.sourcegraph.scip_semanticdb.ScipSemanticdb.processTypedDocument(ScipSemanticdb.java:138)
	at com.sourcegraph.scip_semanticdb.ScipSemanticdb.lambda$runTyped$0(ScipSemanticdb.java:67)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1021)
	at java.util.concurrent.ForkJoinPool$WorkQueue.execLocalTasks(ForkJoinPool.java:1046)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1058)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
time="2022-08-02T11:44:57Z" level=error msg="Process exited with status 1\n"
```

I looked at this a bit to see if I could figure out what's going wrong, however testing on the `2.13.x` branch of Scala I'm also getting the same error in other places:

```
Generating SCIP... [          ] 0%   0 files processedcom.sourcegraph.scip_semanticdb.SignatureFormatterException: failed to format symbol 'local44'
symbol: "local44"
kind: PARAMETER
display_name: "x1"
language: SCALA
signature {
  value_signature {
    tpe {
      type_ref {
        symbol: "scala/Function.untupled(+3).[T1]"
      }
    }
  }
}

Caused by: java.lang.IndexOutOfBoundsException: Index: -1
        at java.util.Collections$EmptyList.get(Collections.java:4456)
        at com.sourcegraph.semanticdb_javac.Semanticdb$TypeRef.getTypeArguments(Semanticdb.java:23534)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatType(SignatureFormatter.java:573)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatValueSignature(SignatureFormatter.java:335)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbolUnsafe(SignatureFormatter.java:69)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbol(SignatureFormatter.java:56)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.processTypedDocument(ScipSemanticdb.java:139)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.lambda$runTyped$0(ScipSemanticdb.java:68)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

So I went a bit deeper down the rabbit hole today. Interesting enough I think there is some collision happening with the `scala` namespace. For example you can minimally reproduce this will the following:

1. Create a `Function.scala` with the following contents exactly:
```scala
package scala

object Function {
  def foo[A](a: A) = a
}
```

2. Run `metac` on this
```sh
metac Function.scala`
```

3. Now run `scip-java index-semanticdb` on the generate semanticdb:
```sh
❯ scip-java index-semanticdb
com.sourcegraph.scip_semanticdb.SignatureFormatterException: failed to format symbol 'scala/Function.foo().'
symbol: "scala/Function.foo()."
kind: METHOD
display_name: "foo"
language: SCALA
signature {
  method_signature {
    type_parameters {
      symlinks: "scala/Function.foo().[A]"
    }
    parameter_lists {
      symlinks: "scala/Function.foo().(a)"
    }
    return_type {
      type_ref {
        symbol: "scala/Function.foo().[A]"
      }
    }
  }
}
access {
  public_access {
  }
}

Caused by: java.lang.IndexOutOfBoundsException: Index: -1
        at java.base/java.util.Collections$EmptyList.get(Collections.java:4586)
        at com.sourcegraph.semanticdb_javac.Semanticdb$TypeRef.getTypeArguments(Semanticdb.java:23534)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatType(SignatureFormatter.java:573)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatTermParameter(SignatureFormatter.java:258)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatScalaParameterList(SignatureFormatter.java:311)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatScalaMethodSignature(SignatureFormatter.java:280)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatMethodSignature(SignatureFormatter.java:211)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbolUnsafe(SignatureFormatter.java:67)
        at com.sourcegraph.scip_semanticdb.SignatureFormatter.formatSymbol(SignatureFormatter.java:56)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.processTypedDocument(ScipSemanticdb.java:138)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.lambda$runTyped$0(ScipSemanticdb.java:67)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:686)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:765)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.runTyped(ScipSemanticdb.java:67)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.run(ScipSemanticdb.java:57)
        at com.sourcegraph.scip_semanticdb.ScipSemanticdb.run(ScipSemanticdb.java:37)
        at com.sourcegraph.scip_java.commands.IndexSemanticdbCommand.run(IndexSemanticdbCommand.scala:90)
        at moped.cli.Command.$anonfun$runAsFuture$1(BaseCommand.scala:14)
        at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
        at scala.util.Try$.apply(Try.scala:210)
        at moped.cli.Command.runAsFuture(BaseCommand.scala:14)
        at moped.cli.Application$.$anonfun$run$7(Application.scala:328)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:434)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

Here's where it gets interesting. Change the package name of the file to `example`:

```scala
package example

object Function {
  def foo[A](a: A) = a
}
```

Now after re-running `metac` again and `scip-java index-semanticdb` you'll see that it succeeds.

--- a/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/SignatureFormatter.java
+++ b/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/SignatureFormatter.java
@@ -23,6 +23,8 @@ public class SignatureFormatter {
 
   private static final Type NOTHING_SYMBOL = typeRef("scala/Nothing#");
   private static final String FUNCTION_SYMBOL_PREFIX = "scala/Function";
+  // Special case scala/Function object to not confluce with Function1 for example
+  private static final String FUNCTION_OBJECT = "scala/Function.";
   private static final String TUPLE_SYMBOL_PREFIX = "scala/Tuple";
   private static final String ARRAY_SYMBOL = "scala/Array#";
   private static final String ENUM_SYMBOL = "java/lang/Enum#";
@@ -557,7 +559,9 @@ public class SignatureFormatter {
           b.append(formatType(typeRef.getTypeArguments(0)));
           b.append("[]");
         }
-      } else if (isScala && typeRef.getSymbol().startsWith(FUNCTION_SYMBOL_PREFIX)) {
+      } else if (isScala
+          && typeRef.getSymbol().startsWith(FUNCTION_SYMBOL_PREFIX)
+          && !typeRef.getSymbol().startsWith(FUNCTION_OBJECT)) {
         int n = typeRef.getTypeArgumentsCount() - 1;
         if (n == 0) {
           // Special-case for Function1[A, B]: don't wrap `A`  in parenthesis like this `(A) => B`

--- a/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/SignatureFormatter.java
+++ b/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/SignatureFormatter.java
@@ -23,7 +23,7 @@ public class SignatureFormatter {
 
   private static final Type NOTHING_SYMBOL = typeRef("scala/Nothing#");
   private static final String FUNCTION_SYMBOL_PREFIX = "scala/Function";
-  // Special case scala/Function object to not confluce with Function1 for example
+  // Special case scala/Function object to not conflict with Function1 for example
   private static final String FUNCTION_OBJECT = "scala/Function.";
   private static final String TUPLE_SYMBOL_PREFIX = "scala/Tuple";
   private static final String ARRAY_SYMBOL = "scala/Array#";
@@ -561,6 +561,7 @@ public class SignatureFormatter {
         }
       } else if (isScala
           && typeRef.getSymbol().startsWith(FUNCTION_SYMBOL_PREFIX)
+          && typeRef.getTypeArgumentsCount() > 0
           && !typeRef.getSymbol().startsWith(FUNCTION_OBJECT)) {
         int n = typeRef.getTypeArgumentsCount() - 1;
         if (n == 0) {

